snapshot/src/main/java/io/camunda/zeebe/snapshots/impl/FileBasedReceivedSnapshot.java

```diff
@@ -96,83 +91,81 @@ private boolean applyInternal(final SnapshotChunk snapshotChunk) throws IOExcept
+  private void checkChunkChecksumIsValid(
+      final SnapshotChunk snapshotChunk, final String snapshotId, final String chunkName)
+      throws SnapshotWriteException {
     final long expectedChecksum = snapshotChunk.getChecksum();
     final long actualChecksum = SnapshotChunkUtil.createChecksum(snapshotChunk.getContent());
 
     if (expectedChecksum != actualChecksum) {
-      LOGGER.warn(
-          "Expected to have checksum {} for snapshot chunk {} ({}), but calculated {}",
-          expectedChecksum,
-          chunkName,
-          snapshotId,
-          actualChecksum);
-      return true;
+      throw new SnapshotWriteException(
+          String.format(
+              "Expected to have checksum %d for snapshot chunk %s (%s), but calculated %d",
+              expectedChecksum, chunkName, snapshotId, actualChecksum));
     }
-    return false;
   }
 
-  private boolean isSnapshotChecksumInvalid(final long currentSnapshotChecksum) {
+  private void checkSnapshotChecksumIsValid(final long currentSnapshotChecksum)
+      throws SnapshotWriteException {
     if (expectedSnapshotChecksum == Long.MIN_VALUE) {
       expectedSnapshotChecksum = currentSnapshotChecksum;
     }
 
     if (expectedSnapshotChecksum != currentSnapshotChecksum) {
-      LOGGER.warn(
-          "Expected snapshot chunk with equal snapshot checksum {}, but got chunk with snapshot checksum {}.",
-          expectedSnapshotChecksum,
-          currentSnapshotChecksum);
-      return true;
+      throw new SnapshotWriteException(
+          String.format(
+              "Expected snapshot chunk with equal snapshot checksum %d, but got chunk with snapshot checksum %d.",
+              expectedSnapshotChecksum, currentSnapshotChecksum));
     }
-    return false;
   }
 
-  private boolean isTotalCountInvalid(final int currentTotalCount) {
+  private void checkTotalCountIsValid(final int currentTotalCount) throws SnapshotWriteException {
     if (expectedTotalCount == Integer.MIN_VALUE) {
       expectedTotalCount = currentTotalCount;
     }
 
     if (expectedTotalCount != currentTotalCount) {
-      LOGGER.warn(
-          "Expected snapshot chunk with equal snapshot total count {}, but got chunk with total count {}.",
-          expectedTotalCount,
-          currentTotalCount);
-      return true;
+      throw new SnapshotWriteException(
+          String.format(
+              "Expected snapshot chunk with equal snapshot total count %d, but got chunk with total count %d.",
+              expectedTotalCount, currentTotalCount));
     }
-    return false;
   }
 
-  private boolean isSnapshotIdInvalid(final String snapshotId) {
+  private void checkSnapshotIdIsValid(final String snapshotId) throws SnapshotWriteException {
     final var receivedSnapshotId = FileBasedSnapshotMetadata.ofFileName(snapshotId);
-    if (receivedSnapshotId.isEmpty()) {
-      return true;
+    if (receivedSnapshotId.isEmpty() || metadata.compareTo(receivedSnapshotId.get()) != 0) {
```

Comment:
Actually, the second condition isn't about format but it's simply the wrong ID, so in that sense the message is confusing. Consider splitting these - it's more verbose, but it saves time when diagnosing issues.